### ISSUE TITLE
USDC Vesper Farming Strategy

### ISSUE THREAD

**Issue by TheRyanMiller** [yearn/yearn-strategies#44]
**Description**: Farm Vesper with USDC. Deposit VSP earnings into vVSP for increased gains.

**Commit (hash)**: [a37070c2bd2b72e785fef31170c659195d01c737](https://github.com/TheRyanMiller/Strategy-Vesper-Farming/commit/a37070c2bd2b72e785fef31170c659195d01c737)

**Tag**: 

**Repo**: https://github.com/TheRyanMiller/Strategy-Vesper-Farming

**Scope**:
- [Strategy.sol](https://github.com/TheRyanMiller/Strategy-Vesper-Farming/blob/master/contracts/Strategy.sol)

**Additional References or Repo** (Link to underlying protocols or relevant documents):
- [Vesper.finance USDC pool](https://app.vesper.finance/vusdc)
- [USDC pool contract](https://etherscan.io/address/0x0C49066C0808Ee8c673553B7cbd99BCC9ABf113d)
- [Vesper.finance VSP pool](https://app.vesper.finance/vvsp)  
- [VSP pool contract](https://etherscan.io/address/0xbA4cFE5741b357FA371b506e5db0774aBFeCf8Fc)  

**Deployed Contract** (etherscan link): TBD

**Review Ongoing By**:
- [ ] Dev 1
- [ ] Dev 2

**Review Completed By**:
- [ ] Dev 1
- [ ] Dev 2

**Comment by Grandthrax** [comment]
https://github.com/TheRyanMiller/Strategy-Vesper-Farming/blob/master/contracts/Strategy.sol#L307
do you mean true?

if(wantValue < callCost) is wayyy too fragile. you should use profitFactor * callCost.

Also you are not converting callCost to want. You need to do that for both tendTrigger and for harvestTrigger

**Comment by Grandthrax** [comment]
https://github.com/TheRyanMiller/Strategy-Vesper-Farming/blob/master/contracts/Strategy.sol#L277

I kind of get what you are doing here. You put this here for tend trigger use. but it does mean that every harvest you will be claiming rewards twice. Once in prepare return and once in adjust position.

**Comment by Grandthrax** [comment]
https://github.com/TheRyanMiller/Strategy-Vesper-Farming/blob/master/contracts/Strategy.sol#L187

In what situation would you every try and withdraw the vsp amount? I dont think it should ever happen as vault never knows about the vsp balance.

But you may trigger this by accident as you dont into account want balance in your calculation.

**Comment by TheRyanMiller** [comment]
> In what situation would you every try and withdraw the vsp amount? I dont think it should ever happen as vault never knows about the vsp balance.

My intent for this design is about security:
- If the strategy gives any public clues as to when it is performing a vVSP harvest (e.g. harvest only when `vVSP balance == xyz`) it could invite profit theft.  
- Therefore, the current mechanic only allows governance / strategist to set the private boolean to `true` when decided that the "big" harvest is profitable.

Please share your thoughts on this if you have any.

**Comment by TheRyanMiller** [comment]
> https://github.com/TheRyanMiller/Strategy-Vesper-Farming/blob/master/contracts/Strategy.sol#L277
> 
> I kind of get what you are doing here. You put this here for tend trigger use. but it does mean that every harvest you will be claiming rewards twice. Once in prepare return and once in adjust position.

Good point. I have added a `claimable()` check to this one in `adjustPosition` to make sure there's anything there first.
But yes, the intent is to be able to claim on either a `harvest` or `tend`.
![image](https://user-images.githubusercontent.com/3812035/111161362-77a47f80-8571-11eb-97dd-32e657bed220.png)

**Comment by TheRyanMiller** [comment]
> https://github.com/TheRyanMiller/Strategy-Vesper-Farming/blob/master/contracts/Strategy.sol#L307
> do you mean true?
> 
> if(wantValue < callCost) is wayyy too fragile. you should use profitFactor * callCost.
> 
> Also you are not converting callCost to want. You need to do that for both tendTrigger and for harvestTrigger

Nice catch. My current plan is to remove this `tendTrigger` override and simply stick to the consistent `_keepVSP` amount rather than spending gas to adjust this amount between harvests. This will produce more consistent returns for depositors.

**Comment by TheRyanMiller** [comment]
Updated issue to link to new commit based on @Grandthrax feedback. For reference old commit is here:
[ab80b89eabdd443f6f60b271814f5622d6af375b](https://github.com/TheRyanMiller/Strategy-Vesper-Farming/commit/ab80b89eabdd443f6f60b271814f5622d6af375b)
